File: common/array/java/ArrayBufferPeer.java
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-5-31  Jed Young  Creation
//

package fan.fgfxArray;

import fan.sys.*;
import java.nio.*;

public class ArrayBufferPeer
{
  private java.nio.Buffer bufferView;
  private ByteBuffer rawData;
  private int offset;
  private int size;

  public static ArrayBufferPeer make(ArrayBuffer self) { return new ArrayBufferPeer(); }

//////////////////////////////////////////////////////////////////////////
// fields
//////////////////////////////////////////////////////////////////////////

  public long size(ArrayBuffer self) { return size; }

  public Endian endian(ArrayBuffer self)
  {
    return (rawData.order() == ByteOrder.BIG_ENDIAN) ? Endian.big : Endian.little;
  }
  public void endian(ArrayBuffer self, Endian v)
  {
    rawData.order(v == Endian.big ? ByteOrder.BIG_ENDIAN : ByteOrder.LITTLE_ENDIAN);
  }

  public NumType type(ArrayBuffer self)
  {
    java.nio.Buffer d = getValue();
    if(d instanceof java.nio.ByteBuffer)
    {
      return NumType.tByte;
    }
    else if (d instanceof java.nio.FloatBuffer)
    {
      return NumType.tFloat;
    }
    else if (d instanceof java.nio.DoubleBuffer)
    {
      return NumType.tDouble;
    }
    else if(d instanceof java.nio.IntBuffer)
    {
      return NumType.tInt;
    }
    else if(d instanceof java.nio.ShortBuffer)
    {
      return NumType.tShort;
    }
    else if(d instanceof java.nio.LongBuffer)
    {
      return NumType.tLong;
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
  }
/*
  public long capacity(ArrayBuffer self) { return getValue().capacity(); }
  public boolean isDirect(ArrayBuffer self) { return getValue().isDirect(); }

  public ArrayBuffer flip(ArrayBuffer self) { getValue().flip(); return self; }
  public long remaining(ArrayBuffer self) { return getValue().remaining(); }
*/
  private void reset()
  {
    getValue().position(this.offset);
    getValue().limit(this.size);
  }

//////////////////////////////////////////////////////////////////////////
// ctor
//////////////////////////////////////////////////////////////////////////

  public static ArrayBuffer allocateDirect(long size, NumType type)
  {
    int capacity = (int)(size * type.size);
    ByteBuffer buf = ByteBuffer.allocateDirect(capacity).order(ByteOrder.nativeOrder());
    buf.clear();

    ArrayBuffer ab = ArrayBuffer.make();
    ab.peer.rawData = buf;
    if (type != NumType.tByte) ab.peer.bufferView = asView(buf, type);
    ab.peer.offset = 0;
    ab.peer.size = (int)size;
    return ab;
  }
  public static ArrayBuffer allocateDirect(long size)
  {
    return allocateDirect(size, NumType.tByte);
  }

  private static java.nio.Buffer asView(ByteBuffer buf, NumType v)
  {
    if (v == NumType.tByte)
    {
      return buf.slice();
    }
    else if (v == NumType.tFloat)
    {
      return buf.asFloatBuffer();
    }
    else if (v == NumType.tDouble)
    {
      return buf.asDoubleBuffer();
    }
    else if(v == NumType.tInt)
    {
      return buf.asIntBuffer();
    }
    else if(v == NumType.tShort)
    {
      return buf.asShortBuffer();
    }
    else if(v == NumType.tLong)
    {
      return buf.asLongBuffer();
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
  }

  public ArrayBuffer createView(ArrayBuffer self, NumType type, long offset, long size)
  {
    if (type(self) != NumType.tByte) throw UnsupportedErr.make("only ByteBuffer can create view");

    this.rawData.position((int)offset);
    this.rawData.limit((int)(offset + size * type.size));

    ArrayBuffer view = ArrayBuffer.make();
    view.peer.rawData = this.rawData;
    view.peer.bufferView = asView(rawData, type);
    view.peer.offset = (int)offset;
    view.peer.size = (int)size;
    reset();

    return view;
  }

  public ArrayBuffer createView(ArrayBuffer self, NumType type, long offset)
  {
    return createView(self, type, offset, this.size / type.size);
  }

  public ArrayBuffer createView(ArrayBuffer self, NumType type)
  {
    return createView(self, type, this.offset);
  }

//////////////////////////////////////////////////////////////////////////
// random read/write
//////////////////////////////////////////////////////////////////////////

  public long getInt(ArrayBuffer self, long index)
  {
    IntBuffer buf = (IntBuffer)getValue();
    return buf.get((int)index);
  }
  public ArrayBuffer set(ArrayBuffer self, long index, long v)
  {
    IntBuffer buf = (IntBuffer)getValue();
    buf.put((int)index, (int)v);
    return self;
  }

  public double getFloat(ArrayBuffer self, long index)
  {
    FloatBuffer buf = (FloatBuffer)getValue();
    return buf.get((int)index);
  }
  public ArrayBuffer setFloat(ArrayBuffer self, long index, double v)
  {
    FloatBuffer buf = (FloatBuffer)getValue();
    buf.put((int)index, (float)v);
    return self;
  }

//////////////////////////////////////////////////////////////////////////
// batch read/write
//////////////////////////////////////////////////////////////////////////

  public ArrayBuffer putFloat(ArrayBuffer self, List list)
  {
    FloatBuffer buf = (FloatBuffer)getValue();
    buf.put(ArrayUtil.toFloatArray(list));
    reset();
    return self;
  }
  public ArrayBuffer putInt(ArrayBuffer self, List list)
  {
    IntBuffer buf = (IntBuffer)getValue();
    buf.put(ArrayUtil.toIntArray(list));
    reset();
    return self;
  }
  public ArrayBuffer putShort(ArrayBuffer self, List list)
  {
    ShortBuffer buf = (ShortBuffer)getValue();
    buf.put(ArrayUtil.toShortArray(list));
    reset();
    return self;
  }

//////////////////////////////////////////////////////////////////////////
// methods
//////////////////////////////////////////////////////////////////////////

  public java.nio.Buffer getValue()
  {
    if (bufferView != null) return bufferView;
    return rawData;
  }

  public void getValue(java.nio.ByteBuffer buf)
  {
    bufferView = null;
    rawData = buf;
  }
}

File: openGl/js/ArrayBufferPeer.js
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-5-29  Jed Young  Creation
//


fan.vaseOpenGl.ArrayBufferPeer = fan.sys.Obj.$extend(fan.sys.Obj);
fan.vaseOpenGl.ArrayBufferPeer.prototype.$ctor = function(self) {}
fan.vaseOpenGl.ArrayBufferPeer.prototype.rawData = null;
fan.vaseOpenGl.ArrayBufferPeer.prototype.bufferView = null;

//////////////////////////////////////////////////////////////////////////
// native
//////////////////////////////////////////////////////////////////////////

fan.vaseOpenGl.ArrayBufferPeer.prototype.m_size;
fan.vaseOpenGl.ArrayBufferPeer.prototype.size = function(self) { return this.m_size; }

fan.vaseOpenGl.ArrayBufferPeer.prototype.m_offset;

fan.vaseOpenGl.ArrayBufferPeer.prototype.m_endian = fan.std.Endian.big;
fan.vaseOpenGl.ArrayBufferPeer.prototype.endian = function(self) { return this.m_endian; }
fan.vaseOpenGl.ArrayBufferPeer.prototype.endian$  = function(self, v) { this.m_endian = v; }
fan.vaseOpenGl.ArrayBufferPeer.prototype.isLittleEndian = function() { return this.m_endian == fan.std.Endian.little; }

fan.vaseOpenGl.ArrayBufferPeer.prototype.type = function(self)
{
  var d = this.bufferView;
  if (d == null) return fan.vaseOpenGl.NumType.m_tByte;
  else if (d instanceof Int8Array) return fan.vaseOpenGl.NumType.m_tByte;
  else if (d instanceof Uint8Array) return fan.vaseOpenGl.NumType.m_tByte;
  else if (d instanceof Int16Array) return fan.vaseOpenGl.NumType.m_tShort;
  else if (d instanceof Uint16Array) return fan.vaseOpenGl.NumType.m_tShort;
  else if (d instanceof Int32Array) return fan.vaseOpenGl.NumType.m_tInt;
  else if (d instanceof Uint32Array) return fan.vaseOpenGl.NumType.m_tInt;
  else if (d instanceof Float32Array) return fan.vaseOpenGl.NumType.m_tFloat;
  else throw fan.sys.Err.make("unknow type");
}

fan.vaseOpenGl.ArrayBufferPeer.prototype.reset = function()
{
  this.m_size = this.bufferView.length;
  this.m_offset = this.bufferView.byteOffset;
}

//////////////////////////////////////////////////////////////////////////
// ctor
//////////////////////////////////////////////////////////////////////////

fan.vaseOpenGl.ArrayBufferPeer.asView = function(rawData, v, offset, size)
{
  var view;
  if (fan.vaseOpenGl.NumType.m_tByte == v) view = new Int8Array(rawData, offset, size);
  else if (fan.vaseOpenGl.NumType.m_tShort == v) view = new Int16Array(rawData, offset, size);
  else if (fan.vaseOpenGl.NumType.m_tInt == v) view = new Int32Array(rawData, offset, size);
  else if (fan.vaseOpenGl.NumType.m_tFloat == v) view = new Float32Array(rawData, offset, size);
  else throw fan.sys.Err.make("unknow type");
  return view;
}
fan.vaseOpenGl.ArrayBufferPeer.prototype.createView = function(self, type, offset, size)
{
  if (this.type(self) != fan.vaseOpenGl.NumType.m_tByte) throw UnsupportedErr.make("only ByteBuffer can create view");

  var view = fan.vaseOpenGl.ArrayBuffer.make();
  var peer = view.peer;
  peer.rawData = this.rawData;
  if (!offset) offset = this.m_offset;
  if (!size) size = this.m_size / type.m_size;
  peer.bufferView = asView(rawData, type, offset, size);
  peer.reset();
  return view;
}

fan.vaseOpenGl.ArrayBufferPeer.allocateDirect = function(size, type)
{
  if (!type) type = fan.vaseOpenGl.NumType.m_tByte;
  var capacity = (size * type.m_size);
  var buf = new ArrayBuffer(capacity);
  var self = fan.vaseOpenGl.ArrayBuffer.make();
  self.peer.rawData = buf;

  if (type == fan.vaseOpenGl.NumType.m_tByte)
  {
    self.peer.bufferView = new Int8Array(buf);
    self.peer.reset();
    return self;
  }
  self.peer.bufferView = fan.vaseOpenGl.ArrayBufferPeer.asView(buf, type, 0, size);
  self.peer.reset();
  return self;
}

//////////////////////////////////////////////////////////////////////////
// random read/write
//////////////////////////////////////////////////////////////////////////

fan.vaseOpenGl.ArrayBufferPeer.prototype.getInt = function(self, index)
{
  if (this.type(self) != fan.vaseOpenGl.NumType.m_tInt) throw UnsupportedErr.make("not Int buffer");
  return this.bufferView[index];
}
fan.vaseOpenGl.ArrayBufferPeer.prototype.setInt = function(self, index, v)
{
  if (this.type(self) != fan.vaseOpenGl.NumType.m_tInt) throw UnsupportedErr.make("not Int buffer");
  this.bufferView.set[index] = v;
  return self;
}
fan.vaseOpenGl.ArrayBufferPeer.prototype.getFloat = function(self, index)
{
  if (this.type(self) != fan.vaseOpenGl.NumType.m_tFloat) throw UnsupportedErr.make("not Float buffer");
  return this.bufferView[index];
}
fan.vaseOpenGl.ArrayBufferPeer.prototype.setFloat = function(self, index, v)
{
  if (this.type(self) != fan.vaseOpenGl.NumType.m_tFloat) throw UnsupportedErr.make("not Float buffer");
  this.bufferView.set[index] = v;
  return self;
}

//////////////////////////////////////////////////////////////////////////
// batch read/write
//////////////////////////////////////////////////////////////////////////

fan.vaseOpenGl.ArrayBufferPeer.prototype.putFloat = function(self, data)
{
  var a = data.toJs();
  if (a.length != this.bufferView.length) {
    a = a.slice(0, this.bufferView.length);
  }
  this.bufferView.set(a);
  return self;
}
fan.vaseOpenGl.ArrayBufferPeer.prototype.putInt = function(self, data)
{
  var a = data.toJs();
  if (a.length != this.bufferView.length) {
    a = a.slice(0, this.bufferView.length);
  }
  this.bufferView.set(a);
  return self;
}
fan.vaseOpenGl.ArrayBufferPeer.prototype.putShort = function(self, data)
{
  var a = data.toJs();
  if (a.length != this.bufferView.length) {
    a = a.slice(0, this.bufferView.length);
  }
  this.bufferView.set(a);
  return self;
}

fan.vaseOpenGl.ArrayBufferPeer.prototype.setFloatArray = function(self, data)
{
  var a = data.m_array;
  if (a.length != this.bufferView.length) {
    a = a.slice(0, this.bufferView.length);
  }
  this.bufferView.set(a);
  return self;
}
fan.vaseOpenGl.ArrayBufferPeer.prototype.setIntArray = function(self, data)
{
  var a = data.m_array;
  if (a.length != this.bufferView.length) {
    a = a.slice(0, this.bufferView.length);
  }
  this.bufferView.set(a);
  return self;
}

//////////////////////////////////////////////////////////////////////////
// methods
//////////////////////////////////////////////////////////////////////////

fan.vaseOpenGl.ArrayBufferPeer.prototype.getValue = function()
{
  return this.bufferView;
}